crates/sovereign-sdk/rollup-interface/src/state_machine/zk/mod.rs

```diff
@@ -249,6 +254,96 @@ pub struct BatchProofCircuitInput<'txs, StateRoot, Witness, Da: DaSpec, Tx: Clon
     pub sequencer_commitments_range: (u32, u32),
 }
 
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize)]
+// Prevent serde from generating spurious trait bounds. The correct serde bounds are already enforced by the
+// StateTransitionFunction, DA, and Zkvm traits.
+#[serde(bound = "StateRoot: Serialize + DeserializeOwned")]
+/// Data required to verify a state transition.
+pub struct BatchProofCircuitInputV2Part1<StateRoot, Da: DaSpec> {
+    /// The state root before the state transition
+    pub initial_state_root: StateRoot,
+    /// The state root after the state transition
+    pub final_state_root: StateRoot,
+    /// The hash before the state transition
+    pub prev_soft_confirmation_hash: [u8; 32],
+    /// The `crate::da::DaData` that are being processed as blobs. Everything that's not `crate::da::DaData::SequencerCommitment` will be ignored.
+    pub da_data: Vec<Da::BlobTransaction>,
+    /// DA block header that the sequencer commitments were found in.
+    pub da_block_header_of_commitments: Da::BlockHeader,
+    /// The inclusion proof for all DA data.
+    pub inclusion_proof: Da::InclusionMultiProof,
+    /// The completeness proof for all DA data.
+    pub completeness_proof: Da::CompletenessProof,
+    /// Pre-proven commitments L2 ranges which also exist in the current L1 `da_data`.
+    pub preproven_commitments: Vec<usize>,
+    /// DA block headers the soft confirmations was constructed on.
+    pub da_block_headers_of_soft_confirmations: VecDeque<Vec<Da::BlockHeader>>,
+    /// The range of sequencer commitments that are being processed.
+    /// The range is inclusive.
+    pub sequencer_commitments_range: (u32, u32),
+}
+
+impl<'txs, StateRoot, Witness, Da, Tx> BatchProofCircuitInput<'txs, StateRoot, Witness, Da, Tx>
+where
+    Da: DaSpec,
+    Tx: Clone,
+    StateRoot: Serialize + DeserializeOwned,
+    Witness: Serialize + DeserializeOwned,
+{
+    /// Into Kumquat expected inputs
+    pub fn into_v2_parts(
+        self,
+    ) -> (
+        BatchProofCircuitInputV2Part1<StateRoot, Da>,
+        BatchProofCircuitInputV2Part2<'txs, Witness, Tx>,
+    ) {
+        assert_eq!(
+            self.soft_confirmations.len(),
+            self.state_transition_witnesses.len()
+        );
+        let mut x = VecDeque::with_capacity(self.soft_confirmations.len());
+
+        for (confirmations, witnesses) in self
+            .soft_confirmations
+            .into_iter()
+            .zip(self.state_transition_witnesses)
+        {
+            assert_eq!(confirmations.len(), witnesses.len());
+
+            let mut v = Vec::with_capacity(confirmations.len());
```

Comment:
Actually constructing Vec from iter.collect uses info about iter len to preallocate a vec of the exact size (for exact size iterators).